### PR TITLE
migrate to typescript

### PR DESCRIPTION
**Pull request by massfords**
## Summary
What does this PR do?
- converts the JS source to TS using the tool from [Stripe](https://github.com/stripe-archive/flow-to-typescript-codemod).

## Details
Why did you make this change? What does it affect?

- exploring what it will take to migrate the remaining repos

## Testing
How can the other reviewers check that your change works?

- there shouldn't be any change to the interfaces or runtime behavior
- I'll try updating an existing script that uses this package

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,38 +12,32 @@
...
   ],
   "scripts": {
-    "build": "yarn run build:babel && yarn run build:flow-copy-source && yarn run build:license-checker",
-    "build:babel": "babel src --out-dir lib --ignore test.js",
-    "build:flow-copy-source": "flow-copy-source --ignore test.js src lib",
+    "build": "yarn run build:license-checker && tsc",
+    "build:flowtypes": "find lib -type f -name '*.d.ts' -exec sh -c 'yarn flowgen --add-flow-header $1 -o ${1%.*.*}.js.flow' _ '{}' \\;",
     "build:license-checker": "license-checker --csv --out var/licenses.csv",
     "clean": "shx rm -rf lib",
+    "flow": "echo this is not a flow project",
```
Comment (massfords) [inline package.json:22]: @jeffsays - there's a workflow script w/ a `yarn flow start` so I left this behind

File: package.json [hunk 2]
```diff
@@ -12,38 +12,32 @@
...
     "git:commit": "git add package.json .yarn/versions && git commit -m v${npm_package_version}",
     "git:commit:publish": "git add package.json .yarn/versions/* && git commit -m v${npm_package_version}",
     "git:commit:version": "git add .yarn/versions/* && git commit -m 'Deferred version bump'",
     "git:push": "git push --follow-tags",
     "git:tag": "git tag -a v${npm_package_version} -m v${npm_package_version}",
     "npm:tag:remove": "yarn npm tag remove ${npm_package_name} $0",
-    "prepack": "yarn run build",
+    "prepack": "yarn run build && yarn run build:flowtypes",
```
Comment (massfords) [inline package.json:29]: defer creation of flow types to the prepack to speed local builds

this is documented in the conversion steps

File: package.json (hunk 1 above)
Comment (jeffsays) [reply package.json:22]: i like it

**Review by jeffsays — APPROVED** [review]
🚀 🚀 🚀 🚀 🚀 🚀

**Review by joshuanapoli — APPROVED** [review]
Thank you!

**Pull request merged**